conans/test/functional/graph_lock/graph_lock_build_requires_test.py

```diff
@@ -0,0 +1,192 @@
+        self.assertEqual(expected, json.loads(t.load("bo.json")))
+
+        # This works at the moment, but we don't know which node it is building
+        t.run("install protobuf/0.1@ --lockfile=conan.lock --build=protobuf "
+              "--lockfile-out=conan.lock")
+        # FIXME: We need a way to speficy contexts or the ID of the node in the lockfile
+        # FIXME: plus maybe the context too, to dissambiguate nodes in both contexts
+
+    def test_build_require_not_removed(self):
+        client = TestClient()
+        client.save({"conanfile.py": GenConanfile()})
+        client.run("create . cmake/1.0@")
+        client.save({"conanfile.py": GenConanfile().with_build_require_plain("cmake/1.0@")})
+        client.run("create . flac/1.0@")
+        client.run("lock create --reference=flac/1.0@ --lockfile-out=conan.lock --build")
+        lock = json.loads(client.load("conan.lock"))
+        flac = lock["graph_lock"]["nodes"]["1"]
+        if client.cache.config.revisions_enabled:
+            ref = "flac/1.0#98ed25e4bb9bc0fdc6d5266afa81f9cf"
+            prev = "405174d701cf8c5478230a92bcc5cf75"
+        else:
+            ref = "flac/1.0"
+            prev = "0"
+        self.assertEqual(flac["ref"], ref)
+        self.assertEqual(flac["package_id"], "5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9")
+        self.assertIsNone(flac.get("prev"))
+
+        client.run("install flac/1.0@ --lockfile=conan.lock --lockfile-out=output.lock")
+        lock = json.loads(client.load("output.lock"))
+        flac = lock["graph_lock"]["nodes"]["1"]
+        self.assertEqual(flac["ref"], ref)
+        self.assertEqual(flac["package_id"], "5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9")
+        self.assertIsNone(flac.get("prev"))
+
+        client.run("install flac/1.0@ --build=flac --lockfile=conan.lock --lockfile-out=output.lock")
+        lock = json.loads(client.load("output.lock"))
+        flac = lock["graph_lock"]["nodes"]["1"]
+        self.assertEqual(flac["ref"], ref)
+        self.assertEqual(flac["package_id"], "5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9")
+        self.assertEqual(flac["prev"], prev)
+
+    def test_multiple_matching_build_require(self):
+        client = TestClient()
+        client.save({"conanfile.py": GenConanfile()})
+        client.run("create . cmake/1.0@")
+        client.run("create . cmake/1.1@")
+        client.save({"conanfile.py": GenConanfile().with_build_require_plain("cmake/1.0")})
+        client.run("create . pkg1/1.0@")
+        client.save({"conanfile.py": GenConanfile().with_build_require_plain("cmake/1.1")
+                                                   .with_require_plain("pkg1/1.0")})
+        client.run("create . pkg2/1.0@")
+        client.run("lock create --reference=pkg2/1.0@ --build --lockfile-out=conan.lock")
+        client.run("install cmake/[>=1.0]@ --lockfile=conan.lock", assert_error=True)
+        self.assertIn("ERROR: Multiple matches in lockfile: 'cmake/1.0', 'cmake/1.1'",
```

Comment:
Should this be an error or just a warning and pick the most updated semver version?

Reply:
In this case I thought it was better to error, as you can narrow the range. If using lockfiles because you want a high determinism, it seems that version range resolution is not the best. I was close to completely forbid using version ranges for this cases, and raising, but I understand it can be convenient, so I left it, as long as it resolves to only one. Note that this ``install <ref> --lockfile=conan.lock`` is typically used with ``--build=<ref>`` to re-build affected packages. So it would be a bad thing if the range can resolve to something different.  Wdyt @czoido?